rust/lance-encoding/src/encodings/physical/binary.rs

```diff
@@ -0,0 +1,421 @@
+            bytes_to_skip, num_bytes
+        );
+
+        // copy target_offsets into dest_buffers[0]
+        let values = target_offsets.values();
+        let byte_slice: &[u8] = unsafe {
+            std::slice::from_raw_parts(
+                values.as_ptr() as *const u8,
+                values.len() * std::mem::size_of::<u32>(),
+            )
+        };
+
+        dest_buffers[0].extend_from_slice(byte_slice);
+        self.bytes_decoder
+            .decode_into(bytes_to_skip, num_bytes, &mut dest_buffers[1..])?;
+
+        // for i in 0..3 {
+        //     println!(
+        //         "length, capacity of buffer: {:?}, {:?} ",
+        //         dest_buffers[i].len(),
+        //         dest_buffers[i].capacity()
+        //     );
+        // }
+
+        Ok(())
+    }
+
+    fn num_buffers(&self) -> u32 {
+        self.bytes_decoder.num_buffers() + 1
+        // 0
+    }
+}
+
+#[derive(Debug)]
+pub struct BinaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    bytes_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl BinaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        bytes_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            bytes_encoder,
+        }
+    }
+}
+
+// Returns indices arrays from string arrays
+// These arrays are padded with a 0 at the beginning
+fn get_indices_from_string_arrays(arrays: &[ArrayRef]) -> Vec<ArrayRef> {
+    let index_arrays: Vec<ArrayRef> = arrays
+        .iter()
+        .filter_map(|arr| {
+            if let Some(string_arr) = arr.as_any().downcast_ref::<StringArray>() {
+                println!("String array: {:?}", string_arr);
+                let offsets = string_arr.value_offsets().to_vec();
+
+                // Not using UInt64, since the range values during decoding expect u32's 
+                let mut new_int_arr = Int32Array::from(offsets);
+                new_int_arr = new_int_arr.slice(1, new_int_arr.len() - 1);
+                // match cast(&new_int_arr, &DataType::UInt64) {
+                //     Ok(res) => Some(Arc::new(res) as ArrayRef),
+                //     Err(_) => panic!("Failed to cast to uint64"),
+                // }
+                Some(Arc::new(new_int_arr) as ArrayRef)
+            } else {
+                panic!("Failed to downcast data to string array");
+            }
+        })
+        .collect();
+
+    index_arrays
+}
+
+fn get_bytes_from_string_arrays(arrays: &[ArrayRef]) -> Vec<ArrayRef> {
+    let byte_arrays: Vec<ArrayRef> = arrays
+        .iter()
+        .filter_map(|arr| {
+            if let Some(string_arr) = arr.as_any().downcast_ref::<StringArray>() {
+                let values = string_arr.values().to_vec();
+                let bytes_arr = Arc::new(UInt8Array::from(values)) as ArrayRef;
+                Some(bytes_arr)
+            } else {
+                panic!("Failed to downcast data to string array");
+            }
```

Comment:
Anytime you have:

```
if let Some(foo) = arr.as_any().downcast_ref():: {

} else {
  panic!(...)
}
```

You should instead use:

```
let foo = arr.as_...();
```

The only time to use `downcast_ref` is when you don't know the type of the array and you are testing if it is type X or not.  In other words, it should look like...

```
if let Some(foo) = arr.as_any().downcast_ref::... {
  // Branch 1
} else {
  // Branch 2 (not panic)
}
```